Enum functions

sinansevgi opened the pull request:
**What does this do?**

- Clones of Ruby Enumerable Functions

**Description of Task to be completed?**

- In this project, We have recreated our own versions of enumerable methods.

**How should this be manually tested?**

- Clone this repo https://github.com/sinansevgi/enum_methods/
- Navigate to project directory.
- Open the terminal and run ruby my_methods.rb

sinansevgi committed:
my_each, my_each_with_index and my_select functions created.

rukundoeric committed:
Add my-all,my-any,my-none and my-count functions

sinansevgi committed:
Feature my_map, my_inject

my_map, my_inject functions created.

rukundoeric committed:
Add ReadMe

rukundoeric committed:
Merge branch 'enum_functions' of https://github.com/sinansevgi/enum_methods into enum_functions

rukundoeric committed:
Modify ReadMe

voscarmv commented on my_methods.rb:
```diff
@@ -0,0 +1,67 @@
+def my_any(array)
+  for i in 0...array.length
+    if(yield array[i])
+      return true
+    end
+  end
+  return false
+end
+
+def my_none(array)
+  for i in 0...array.length
+    if(yield array[i])
+      return false
+    end
+  end
+  return true
+end
+
+def my_count(array)
+  count = 0
+  for i in 1..array.length
+    count += 1
+  end
+  count
+end
+
+def my_map(array, funct)
+  result_array = []
+  for item in array
+    result_array.push(method(funct).call(item))
+  end
+  return result_array
+end
+
+def my_inject(array, result, funct)
+  for item in array
+    result = method(funct).call(result, item)
+  end
+  return result
+end
```
All of your methods are supposed to be wrapped around `def Enumerable` in order to extend Ruby's native `Enumerable` class. Like this:

```
def Enumerable
   # ...your methods here...
end
```

voscarmv reviewed (changes requested):
# Nice work so far, team. 👍 
## However, there are some changes required 👀 🧰 🔨 🔧 

- Before we can test any of your methods, please remember that you're supposed to extend the `Enumerables` library. Please read my code comments on how to do this and fix your code.
- Also, please make sure your project has the proper linter configuration: https://github.com/microverseinc/linters-config/tree/master/ruby

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

sinansevgi committed:
.yml file fixed

Rspec removed.

Bug fix

Linter errors

Bug fix

Enumerable class definition fixed

Linter configuration added.

Functions nested with class.

sinansevgi committed:
Linter configuration added.

Functions nested with class.

voscarmv commented:
Hi, sorry, I must make a correction. You should add `module Enumerable` instead of `def Enumerable`

sinansevgi committed:
Bug fix

Enumerable class definition fixed

sinansevgi committed:
Bug fix

Linter errors

sinansevgi committed:
Rspec removed.

sinansevgi committed:
.yml file fixed

sinansevgi committed:
Bug fix string literal

Unnecessary frozen string literal removed

sinansevgi committed:
Bug fix string literal

Unnecessary frozen string literal removed

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+# rubocop:disable all
```
- [x] Enable Rubocop and run `rubocop` on the console. There are issues that need to be addressed before submitting the code for a re-review.

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+    for item in array
+      yield item
+    end
+  end
```
- [x] `my_each` is not working as the original `each` method, also is not passing all test cases.

### `each` is an Enumerator that has these characteristics: 
- Returns an Enumerator if no block is given.
```
[1, 2, 3].each 
# #<Enumerator: [1, 2, 3]:each>
````
- It does not change or modify the original array/range

- it returns the array/range/hash itself if a block is given.
```
[1, 2, 3].each { |num| num }
# [1, 2, 3]
(1..3).each { |num| num * 2 }
# 1..3
````
- it applies the block for each element inside the array/range/hash
```
[1, 2, 3].each { |num| puts num }
# 1
# 2
# 3
# => [1, 2, 3]
```
[Enumerator#each](https://ruby-doc.org/core-2.7.2/Enumerator.html#method-i-each)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+    for i in (0...array.length)
+      yield i, array[i]
+    end
+  end
```
- [x] `my_each_with_index` is not working as the original `each_with_index` method, also is not passing all test cases.

### `each_with_index` is an Enumerable that has these characteristics: 
- Returns an Enumerator if no block is given.
```
[1, 2, 3].each_with_index
# #<Enumerator: [1, 2, 3]:each_with_index>
````
- It does not change or modify the original array/range

- it returns the array/range/hash itself if a block is given.
```
[1, 2, 3].each_with_index { |num, idx| num }
# [1, 2, 3]
(1..3).each { |num, idx| num * idx }
# 1..3
````
- it applies the block for each element inside the array/range/hash
```
[1, 2, 3].each_with_index { |num, idx| puts num, idx }
# 1
# 0
# 2
# 1
# 3
# 2
# => [1, 2, 3]
```
[Enumerator#each](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-each_with_index)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+    end
+    return true
+  end
```
- [x] `my_all` is not working as the original `all?` method, also is not passing all test cases. 

- [x] the method name should have `?` question mark appended to the end of the method name.

### `all?` is an Enumerable that has these characteristics:
- Returns `true` if all elements are not false.
```
[1, 'string', true].all?
# true
```
- Returns `false` if it finds at least one element that is not true
```
[1, 'string', nil].all?
# false
```
- it doesn't change the original array/range/hash
- it can accept as arguments a Class, a RegEx, or a pattern, and returns true or false depending if the values in the array/range/hash match the condition
- it can use a block but it is not necessary to make the method work

[Enumerable#all?](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-all-3F)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+    end
+    return false
+  end
```
- [x] `my_any` is not working as the original `any?` method, also is not passing all test cases. 

- [x] the method name should have `?` question mark appended to the end of the method name.

### `any?` is an Enumerable that has these characteristics:
- Returns `true` if at least one element is not false.
```
[1, false, nil].any?
# true
```
- Returns `false` if all elements are false.
```
[nil, false, nil].any?
# false
```
- it doesn't change the original array/range/hash
- it can accept as arguments a Class, a RegEx, or a pattern, and returns true or false depending if the values in the array/range/hash match the condition
- it can use a block but it is not necessary to make the method work

[Enumerable#any?](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-any-3F)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+    end
+    return true
+  end
```
- [x] `my_none` is not working as the original `none?` method, also is not passing all test cases. 

- [x] the method name should have `?` question mark appended to the end of the method name.

### `none?` is an Enumerable that has these characteristics:
- Returns `true` if all elements are false.
```
[nil, nil, nil].none?
# true
```
- Returns `false` if it finds at least one element that is true
```
[1, 'string', nil].none?
# false
```
- it doesn't change the original array/range/hash
- it can accept as arguments a Class, a RegEx, or a pattern, and returns true or false depending if the values in the array/range/hash match the condition
- it can use a block but it is not necessary to make the method work

[Enumerable#none?](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-none-3F)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+      count += 1
+    end
+    count
+  end
```
- [x] `my_count` is not working as the original `count` method, also is not passing all test cases. 

### `count` is an Enumerable that has these characteristics:
- Returns the count of elements inside an array/range/hash.
- When a block is given it returns the count of elements that match the condition
```
[1, 2, 3].count { |num| num > 2 }
# 1
```
[Enumerable#count](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-count)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+      result_array.push(method(funct).call(item))
+    end
+    return result_array
+  end
```
- [x] `my_map` is not working as the original `map` method, also is not passing all test cases.

### `map` is an Enumerable that has these characteristics:
- Returns an Enumerator if no block is given.
```
[1, 2, 3].map
# #<Enumerator: [1, 2, 3]:map>
```
- Returns a **new array** with the results of running a block
```
[1, 2, 3].map { |num| num * 2 }
# [2, 4, 6]
```
- doesn't change/modify the original array/range/hash
- can accept a proc as an argument
```
['1', '2', '3'].map(&:to_i)
# [1, 2, 3]
```
[Enumerable#map](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-map)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+    for i in array
+      yield i if i != filter
+    end
+  end
```
- [x] `my_select` is not working as the original `select` method, also is not passing all test cases.

### `select` is an Enumerable that has these characteristics:
- Returns an Enumerator if no block is given.
```
[1, 2, 3].select
# #<Enumerator: [1, 2, 3]:select>
```
- Returns a **new array** with the results that match the condition in the block
```
[1, 2, 3].select { |num| num > 2 }
# [3]
```
- doesn't change/modify the original array/range/hash

[Enumerable#select](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-select)

jubaan commented on my_methods.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+    return result
+  end
+end
```
- [x] `my_inject` is not working as the original `inject` method, also is not passing all test cases.

### `inject`` is an Enumerable that has these characteristics:
- Returns a `LocalJumpError` when no block is given and no symbol is passed as an argument.
```
[1, 2, 3].inject
# LocalJumpError (no block given)
```

- it does not modify/change the original array/range/hash.

- it can accept an initial value and/or a symbol as an argument.
```
[1, 2, 3].inject(2, :*)
# 12
```

[Enumerable#inject](https://ruby-doc.org/core-2.7.2/Enumerable.html#method-i-inject)

jubaan reviewed (changes requested):
# Status: Changes Required :recycle: 

## :wave: :wave: Hello @sinansevgi and @rukundoeric  . Thanks for submitting you project!

Great thing this project has and has done well:
:100: Extended the `Enumerable` with a `module` 
:100: Descriptive README
:100: Descriptive PR

There are some issues that need to be addressed before we can move forward. Please review the list below.

## Required changes :wrench: 
### [Mandatory] :red_circle: 

- [x] Rubocop linter should not be disabled. See inline comment.

- [x] All methods throw an `ArgumentError`. This error is caused by the way the methods are declared. Some methods may need arguments or might use arguments to do operations, and others might not need them at all. To work around this issue go through the Ruby documentation and analyze how each of the `Enumerator`/`Enumerable` works with the examples given. [Enumerator](https://ruby-doc.org/core-2.7.2/Enumerator.html) | [Enumerable](https://ruby-doc.org/core-2.7.2/Enumerable.html)

- [x] should have a working `my_each`. See inline comment.

- [x] should have a working `my_each_with_index`. See inline comment.

- [x] should have a working `my_select`. See inline comment.

- [x] should have a working `my_all?`. See inline comment. 

- [x] should have a working `my_any?`. See inline comment. 

- [x] should have a working `my_none?`. See inline comment. 

- [x] should have a working `my_count`. See inline comment. 

- [x] should have a working `my_map`. See inline comment.

- [x] should have a working `my_inject`. See inline comment.

- [x] The project is missing the `multiply_els` method

### [OPTIONAL]  :yellow_circle:  

- [x] Twitter Link for @sinansevgi is broken. Clicking on the link redirects to an account that doesn't exist. Fix the link to improve the professionalism of the README.

- [x] It would be great to see in the readme not only the methods but also some outputs of the methods that show the reader what to expect when using it.

- [x] `my_all`, `my_any`, and `my_none` are methods that return boolean values, Ruby style conventions suggest to append to the method name a `?` question mark. This improves the readability of the code. Also helps the programmer understand and know in advance what to expect. [Predicate Methods Suffix](https://rubystyle.guide/#bool-methods-qmark)

## Keep up the great work :muscle: and Happy Coding :smiley: :computer:!!

rukundoeric committed:
Implement Feedback

rukundoeric committed:
Implement Feedback

sinansevgi committed:
Live version link updated

Update README.md

Error fix
Module length suppressed.
Linter errors fixed.

Bug fix
my_inject function fixed.

sinansevgi committed:
Bug fix

my_inject function fixed.

sinansevgi committed:
Error fix

Linter errors fixed.

sinansevgi committed:
Error fix

Module length suppressed.

sinansevgi committed:
Readme file updated.

sinansevgi committed:
Update README.md

Commit:
Live version link updated/

blackpintz commented on my_methods.rb:
```diff
@@ -0,0 +1,199 @@
+    self
+  end
+
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    array = self
+    array = array.to_a if array.class == Range
+    array = array.values if array.class == Hash
+
+    for i in (0...array.length)
+      yield array[i], i
+    end
+  end
```
The `my_each_with_index` method for hash values does not give the correct output.

```
h = {foo: 0, bar: 1, baz: 2}

#each_with_index

hash = Hash.new
h.each_with_index { |item, index|
  hash[item] = index
}

print hash  #=> {[:foo, 0]=>0, [:bar, 1]=>1, [:baz, 2]=>2}

#my_each_with_index

hash = Hash.new
h.my_each_with_index { |item, index|
  hash[item] = index
}

print hash #=> {0=>0, 1=>1, 2=>2}
```

blackpintz commented on my_methods.rb:
```diff
@@ -0,0 +1,199 @@
+    res
+  end
+
+  def my_all?(arg = nil)
+    return false if !block_given? && !arg
+
+    array = self
+    array = array.to_a if array.class == Range
+    array = array.values if array.class == Hash
+
+    return true if array.empty?
+
+    for i in 0...array.length
+      if arg
+        if arg.is_a? Module or arg.is_a? Class
+          return false unless array[i].is_a?(arg)
+        elsif arg.class == Regexp
+          return false if array[i].match(arg).nil?
+        else return false unless array[i] != arg
+        end
+      else return false unless yield array[i]
+      end
+    end
+
+    true
+  end
```
Fix the following for the` my_all?` method.
```

print [].all?  #=> true
print [].my_all?  #=> false

print (0..5).all?  #=> true
print (0..5).my_all?  #=> false
```

blackpintz commented on my_methods.rb:
```diff
@@ -0,0 +1,199 @@
+    true
+  end
+
+  def my_any?(arg = nil)
+    array = self
+    array = array.to_a if array.class == Range
+    array = array.values if array.class == Hash
+
+    return false if array.empty?
+    return true if array.include? true
+
+    for i in 0...array.length
+      if arg
+        if arg.is_a? Module or arg.is_a? Class
+          return true if array[i].is_a?(arg)
+        elsif arg.class == Regexp
+          return true if array[i].match(arg).nil?
+        else return true unless array[i] != arg
+        end
+      else return true if yield array[i]
+      end
+    end
+    false
+  end
```
Fix the following for `my_any?` method

```
puts (0..5).any?   #=> true
puts (0..5).my_any?  #=> my_methods.rb:89:in `block in my_any?': no block given (yield) (LocalJumpError)

#  in the case below it checks if any word has a letter d.
puts %w[ant bear cat].any?(/d/)    #=> false
puts %w[ant bear cat].my_any?(/d/)   #=> true
```

blackpintz commented on my_methods.rb:
```diff
@@ -0,0 +1,199 @@
+    false
+  end
+
+  def my_none?(arg = nil)
+    return true if !block_given? && !arg
+
+    array = self
+    array = array.to_a if array.class == Range
+    array = array.values if array.class == Hash
+
+    for i in 0...array.length
+      if arg
+        if arg.is_a? Module or arg.is_a? Class
+          return false if array[i].is_a?(arg)
+        elsif arg.class == Regexp
+          return false if array[i].match(arg).nil?
+        else return false unless array[i] != arg
+        end
+      else return false if yield array[i]
+      end
+    end
+
+    true
+  end
```
Fix the following for `my_none?` method

```
# It should return true since none of the words contain letter d.
puts %w{ant bear cat}.none?(/d/)    #=> true
puts %w{ant bear cat}.my_none?(/d/)  #=> false

puts [nil, false, true].none?          #=> false
puts [nil, false, true].my_none?   #=> true

puts (0..4).none?    #=> false
puts (0..4).my_none?  #=> true
```

blackpintz reviewed (changes requested):
# Status: Changes Required

## Hello team,

**Great job implementing the changes. There are still a few fixes needed. Please check the comments below (all are mandatory).** 

### Happy coding 😃.

**PS: You can reach me on slack (Rose Wanjohi).**

rukundoeric committed:
Implement Feedback

rukundoeric committed:
Implement feedback about my_none, my_each_with_index and my_all

rukundoeric committed:
Implement Feedback

santiagorodriguezbermudez commented on my_methods.rb:
```diff
@@ -0,0 +1,220 @@
+    self
+  end
+
+  def my_each_with_index
```
When I compare the original method to yours I get some different results:

- Perhaps work with lines 26 and 27 to change the conditionals when the user passes an array as self.

santiagorodriguezbermudez commented on my_methods.rb:
```diff
@@ -0,0 +1,220 @@
+    false
+  end
+
+  def my_none?(arg = nil)
```
Review two scenarios:

1. When the user passes a pattern other than a Class or Regex, the method is returning false when none of the collection matches the pattern. It should return true.

2. When the user passes in a Regex as an argument it should return false if any of the collection matches the Regex. Your method returns true. Try using the match? method for comparing regex expressions.

santiagorodriguezbermudez commented on my_methods.rb:
```diff
@@ -0,0 +1,220 @@
+  end
+
+  def my_count(arg = nil)
+    return length if !block_given? && !arg
```
Length is not defined here when working with a Range. Be sure to fix this.

santiagorodriguezbermudez reviewed (changes requested):
# Status: Changes Required ♻️🚦

Hi @sinansevgi and @rukundoeric! Hope all is well. 

Good work so far. You're almost there. 

Tests on your methods are not passing. I have the following errors:
- [x] my_each_with_index test is failing. As the original method, it should return the array itself after calling the given block once for each element. 
![Screen Shot 2020-10-22 at 11 24 55 AM](https://user-images.githubusercontent.com/5817056/96901702-37752100-1459-11eb-83f5-d49d938e7317.png)
- [x] my_all? When a pattern other than Regex or a Class is given it should return true if all of the collection matches the pattern. Currently, your method returns false.
- [x] my_none? When a Regex is passed as an argument it should return false if any of the collection matches the Regex. Currently, your method returns true.
- [x] my_count should return the number of items in enum. When I run the method I get an undefined error for the length method. Please review this. 

Moreover, your latest commits are repetitive. Please review them and make sure each commit is unique and descriptive about the changes you made. Merge the last commits into one. This will keep your commits and repo professional. 

*Please check the inline comments for more information*. 

👨🏻‍💻 👩‍💻 Happy Coding, 

If you have any further questions or comments, please reach out through slack @santiagorodriguezbermudez.

rukundoeric committed:
Implement feedback about my_none, my_each_with_index and my_all

acushlakoncept reviewed (changes requested):
# Status: Changes Required ♻️

Hello @rukundoeric and @sinansevgi, Both of you have done well on this project.
Let's make a few changes, and we will be done.

- [x] Your `#my_any?` method should return true if at least one of the collections is not false or nil when no block or argument is given. Eg. `[nil, false, true, []].my_any?` should return true

- [x] `#my_none?` method should return false only if one of the collection matches the pattern when a pattern other than Regex or a Class is given. Eg. `[5, "door", "rod", "blade"].my_none?(5)` should return `false`

- [x] You missed the `#multiply_els` method in your methods. Kindly add it as one of your methods. 
<img width="1028" alt="Screen Shot 2020-10-23 at 1 14 06 AM" src="https://user-images.githubusercontent.com/5264771/96942466-1e0dbc80-14cd-11eb-96a9-9c3971d778eb.png">

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

sinansevgi committed:
Merge remote-tracking branch 'origin/enum_functions' into enum_functions

rukundoeric committed:
Error fix on my_none & my_any

rukundoeric committed:
Add multiply_els

euqueme commented on my_methods.rb:
```diff
@@ -0,0 +1,230 @@
+    false
+  end
+
+  def my_none?(arg = nil)
+    array = self
+    array = array.to_a if array.class == Range
+    array = array.values if array.class == Hash
+
+    if !block_given? && !arg
+      res = array.my_all? do |x|
+        !(x != false and !x.nil?)
+      end
+
+      return true if res
+
+      return false
+    end
+
+    for i in 0...array.length
+      if arg
+        if arg.is_a? Module or arg.is_a? Class
+          return false if array[i].is_a?(arg)
+        elsif arg.class == Regexp
+          return false unless array[i].match(arg).nil?
+        else
+          return false if array[i] == arg
+        end
+      else return false if yield array[i]
+      end
+    end
+
+    true
+  end
```
### [OPTIONAL]

- [ ] I think we could get the block of my_none? from my_any? we should only do something like !my_any? for my_none? to work. For example:

```ruby
def my_none?(arg = nil, &block)
  !my_any?(arg, &block)
end
```

To make this work we have to use block.call(element) instead of yield(element) in my_any?

euqueme reviewed (approved):
> ### Pull Request Reviewed by Maria Eugenia Queme P. (TSE)

### Hello @sinansevgi and @acushlakoncept :balloon: :lollipop: 

### Thank you for doing all the requested changes :1st_place_medal: :1st_place_medal: your Enumerable methods are running smoothly :sparkles: and I'm very happy to inform you to :drum: :drum: :drum: merge this branch and move on its a yes for me Congratulations!! :tada: :tada: :tada: 

### Please :pray:  keep up the good work :clinking_glasses: 

PROJECT STATUS: APPROVED :heavy_check_mark:
